SAK-32035 : Replace courier service in chat with some other service

frasese opened the pull request:
More info in JIRA : https://jira.sakaiproject.org/browse/SAK-32035

mpellicer commented on chat/chat-api/api/src/java/org/sakaiproject/chat2/model/MessageDateString.java:
```diff
@@ -0,0 +1,12 @@
+	private String pdate;
+	private String ptime;
+	private String pid;
+}
```
Please add new line at EOF

steveswinsburg replied:
What's the P prefix?

frasese replied:
I don't know... these are the same names as used in the old chat. Maybe we can change them for  more descriptive ones.

mpellicer commented on chat/chat-tool/tool/src/bundle/chat.properties:
```diff
@@ -157,6 +157,8 @@ add_channel_title=Add Room
 message_display=Message Display
 
 messages_shown_total=Showing {0} messages out of {1}
+messages_unreaded=There are {0} unreaded messages
```
There are {0} unread messages

mpellicer commented on chat/chat-tool/tool/src/bundle/chat.properties:
```diff
@@ -157,6 +157,8 @@ add_channel_title=Add Room
 
 messages_shown_total=Showing {0} messages out of {1}
+messages_unreaded=There are {0} unreaded messages
+unreaded=Unreaded messages
```
Unread messages

mpellicer commented on chat/chat-tool/tool/src/webapp/css/chat.css:
```diff
@@ -0,0 +1,139 @@
+/* sakai chat tool CSS */
```
This classes are already in Morpheus, i think this file is not needed.

mpellicer commented on chat/pom.xml:
```diff
@@ -32,6 +32,17 @@
+          <groupId>org.jgroups</groupId>
+          <artifactId>jgroups</artifactId>
+          <version>3.6.13.Final</version>
+          <!--version>4.0.3.Final</version-->
```
Remove this comment please.

mpellicer commented:
LGTM, terrifying work @frasese & @josecebe  !

frasese commented on chat/chat-api/api/src/java/org/sakaiproject/chat2/model/TransferableChatMessage.java:
```diff
@@ -0,0 +1,95 @@
+	public TransferableChatMessage(MessageType type, String id, String owner, String siteId, String channelId, String content) {
+		this(type, id, owner, siteId, channelId, content, (new Date()).getTime());
+	}
+	public TransferableChatMessage(MessageType type, String id, String owner, String siteId, String channelId, String content, long timestamp) {
```
Yes, it is used to sort messages before send them to the client. So we will be sure the order is right.

ern replied:
my comment was about that the timestamp param is passed into the constructor but it is never used since you always construct a new Date object.

frasese commented on chat/pom.xml:
```diff
@@ -32,6 +32,16 @@
+      </dependency>
+      <dependency>
+          <groupId>com.google.guava</groupId>
+          <artifactId>guava</artifactId>
```
Mmmmm... as I can see, there is not any guava in shared. Other tools/components using it (like kernel, dashboard or search), are loading it into their component folder.
We can move the library to shared, but then, we need to update all other components. Maybe is better do this in another PR.

ern replied:
ok yeah I though it was already in shared ;)

frasese commented on chat/chat-impl/impl/src/webapp/WEB-INF/components.xml:
```diff
@@ -21,8 +21,18 @@
+				<property name="sessionManager" ref="org.sakaiproject.tool.api.SessionManager" />
+				<property name="usageSessionService" ref="org.sakaiproject.event.api.UsageSessionService"/>
+				<property name="formattedText" ref="org.sakaiproject.util.api.FormattedText"/>
+				<property name="timeService" ref="org.sakaiproject.time.api.TimeService"/>
```
Yeah... I have the same improvement in my blackboard with pending tasks. I'm going to take a look at this...

frasese commented:
Updated according previous comments :
- Removed TimeService (uses java8 time and preferences service instead)
- Using lombok's EqualsAndHashcode
- Renamed pdate, ptime, pid with localizedDate, localizedTime, dateID

frasese commented on chat/chat-impl/impl/src/java/org/sakaiproject/chat2/model/impl/ChatManagerImpl.java:
```diff
@@ -123,17 +171,61 @@ protected void init() throws Exception
+                    logger.info("No 'chat.cluster.channel' specified in sakai.properties. JGroups will not be used and chat messages will not be replicated."); 
+                }
+            } catch (Exception e) {
+                logger.error("Error creating JGroups channel. Chat messages will now NOT BE KEPT IN SYNC", e);
```
done

ern commented on chat/chat-api/api/src/java/org/sakaiproject/chat2/model/TransferableChatMessage.java:
```diff
@@ -0,0 +1,95 @@
+	public String siteId;
+	public String channelId;
+	public String content;
+	public long timestamp;
```
Is there any need for TZ to be associated with this time instance?

frasese replied:
At this moment the timestamp is used locally in each cluster node. It represents the moment a message:
- is created locally
- is received via jGroups
I've been testing to send a ZonedDateTime with the moment each message is created in its own node, but if the nodes' clock are not exactly synchronized, this process produces unexpected results.

frasese replied:
Maybe it's a good idea change the ChatMessage.messageDate property from java.util.Date to java.time.Instant (or java.time.ZonedDateTime) to avoid problems saving in DB when cluster nodes are in different Timezones. But seems that, hibernate 4 does not support well these types.

We should move this to a different PR.

ern replied:
Correct just use java.util.Date and then convert to java 8 time for now. H5 supports the new java 8 time types.

ern commented on chat/chat-api/api/src/java/org/sakaiproject/chat2/model/TransferableChatMessage.java:
```diff
@@ -0,0 +1,95 @@
+		out.writeObject(timestamp);
+	}
+
+	public void readObject(java.io.ObjectInputStream in) throws IOException, ClassNotFoundException {
```
any reason to use an import here?

ern commented:
Just a few more minor questions/suggestions and we can get this one in! Again big thanks to @frasese and @josecebe

frasese commented:
Anything else I can do here?
